### PR TITLE
Compound templates

### PR DESCRIPTION
**Pull request by christian-casecommons**
Series of updates to allow for broadened use of compound templates (see https://github.com/Casecommons/cbp-authentication-api-aws/blob/master/templates/stack.yml.j2) 
```
### Version 1.0.0

- **ENHANCEMENT** : Updated autoscaling group template `./templates/asg.yml.j2`
    * Add parameter for custom ecs cluster
    * Updated UserData to deal with custom ecs cluster
    * Added AmazonEC2ContainerServiceforEC2Role to autoscaling iam role
    * Added conditional checks for custom ecs cluster

- **ENHANCEMENT** : Updated application load balancing template `./templates/alb.yml.j2`
    * Added defaults and conditionals for dns domain and host name in case that those are registered separately in `./templates/dns.yml.j2`
    * Added defaults and conditionals for target group in the case that those are registered separetely in `./templates/alb_target.yml.j2`
    * Added additional healthchecks/alarms for target group

- **ENHANCEMENT** : Updated DNS template `./templates/dns.yml.j2`
    * Allow for passing mapping when defining dns records

- **ENHANCEMENT** : Updated RDS template `./templates/dns.yml.j2`
    * Added defaults and conditionals for dns domain and host name in case that those are registered separately in `./templates/dns.yml.j2`
    * Added default of postgres for database type

- **ENHANCEMENT** : Added thin-slice template to define security groups `./templates/security_group.yml.j2`
- **ENHANCEMENT** : Added thin-slice template to define ECS cluster resources, healthchecks and alarms `./templates/ecs.yml.j2`
- **ENHANCEMENT** : Added thin-slice template to encapsulate use of kms decrypt lambda functionality `./templates/decrypt.yml.j2`
- **ENHANCEMENT** : Added thin-slice template to define ECS service and task resources, healthchecks and alarms in  `./templates/ecs_service.yml.j2`
- **ENHANCEMENT** : Added thin-slice template to encapsulate usage of generic lambda function `./templates/lambda.yml.j2`
- **ENHANCEMENT** : Added ansible filter plugin to determine current timestamp `./filter_plugins/timestamp.py`
```

### CODE REVIEW

File: filter_plugins/timestamp.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from time import time
+
+class FilterModule(object):
+  ''' Returns dictionary of items with keyed overrides that override properties with a matching selector'''
```
Comment (mixja) [inline filter_plugins/timestamp.py]: Please update comment to reflect this returns a timestamp

File: templates/asg.yml.j2 [hunk 2]
```diff
@@ -356,26 +383,30 @@ Resources:
...
           Fn::If:
             - ProxyEnabledCondition
-            - Fn::Sub: |
-                #!/bin/bash
-                trap 'kill $(jobs -p)' EXIT
-                tail --retry -F /var/log/cfn-init.log >> /dev/console &
-                /opt/aws/bin/cfn-init -v --stack ${AWS::StackName} \
-                  --resource ${Name}LaunchConfiguration --region ${AWS::Region} \
+            -
+              Fn::Join: ["\n", [
```
Comment (mixja) [inline templates/asg.yml.j2]: There's no reason to use Fn::Join here - Fn::Sub is much easier to read.

You only need to use Fn::Joins typically when you have Fn::ImportValue expressions.

File: templates/asg.yml.j2 [hunk 3]
```diff
@@ -501,11 +530,13 @@ Resources:
               CidrIp: 0.0.0.0/0
       Tags:
         - Key: Name
-          Value: 
+          Value:
             Fn::Sub: ${AWS::StackName}-${Name}-sg
   InstanceRole:
     Type: AWS::IAM::Role
     Properties:
+      ManagedPolicyArns:
```
Comment (mixja) [inline templates/asg.yml.j2]: This managed policy negates the IAM policy below - please see https://github.com/docker-production-aws/microtrader-deploy/blob/final/templates/stack.yml.j2#L1619-L1642 for full list of permissions required.

File: templates/lambda.yml.j2 [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+      - Ref: Version
+      - latest
+  AlwaysRunCondition:
+    Fn::Equals:
+      - Ref: AlwaysRun
+      - "True"
+
+
+Resources:
+  Caller:
```
Comment (mixja) [inline templates/lambda.yml.j2]: What's the use case for running on every stack run?

File: filter_plugins/timestamp.py (hunk 1 above)
Comment (christian-casecommons) [reply filter_plugins/timestamp.py]: good catch

File: templates/lambda.yml.j2 (hunk 4 above)
Comment (christian-casecommons) [reply templates/lambda.yml.j2]: Originally, this was intended for blue/green deploys, where a lambda function was necessary to access state, there being which env is current production. 
https://github.com/Casecommons/lambda-router-tasks/pull/1
https://github.com/Casecommons/lambda-router-tasks/blob/1fd3f5433af3469ef997df19daafdb08fc725b52/src/main.py#L50

Other cases include running migrations for database backed services.

**Comment by mixja** [comment]
Closing as inactive

**Pull request closed without merging**